executors.md

```diff
@@ -0,0 +1,179 @@
+Reform], but also makes some improvements to the `current_thread` design.
+
+# Proposed design
+
+## The core executor abstraction: `Spawn`
+
+First, in `futures-core` we provide a basic definition of a general-purpose
+executor:
+
+```rust
+trait Spawn {
+    fn spawn(&self, f: Box<Future<Item = (), Error = ()> + Send>);
+}
+```
+
+The `Spawn` trait is quite straightforward (some alternatives and tradeoffs are
+discussed in the next section), and crucially is object-safe.
+
+We then build in an executor to the task context, stored internally as a trait
+object, which allows us to provide the following methods:
+
+```rust
+impl task::Context {
+    // Spawn onto the current default executor
+    fn spawn<F>(&self, F)
+        where F: Future<Item = (), Error = ()> + Send + 'static;
+
+    // Run the given closure with a context using a new default executor
+    fn with_spawn<S, F, R>(&mut self, spawn: S, f: F) -> R
+        where S: Spawn, F: FnOnce(&mut task::Context) -> R;
+}
+```
+
+With those APIs in place, we've achieved two of our goals: it's possible for any
+future to spawn a new task, and to exert fine-grained control over generic task
+spawning within a sub-future.
+
+## Built-in executors
+
+In the `futures-executor` crate, we then add two executors: `ThreadPool` and
+`LocalPool`, providing multi-threaded and single-threaded execution,
+respectively.
+
+### Multi-threaded execution: `ThreadPool`
+
+The `ThreadPool` executor works basically like `CpuPool` today:
+
+```rust
+struct ThreadPool { ... }
+impl Spawn for ThreadPool { ... }
+
+impl ThreadPool {
+    // sets up a pool with the default number of threads
+    fn new() -> ThreadPool;
+
+    // starts up the worker threads
+    fn run();
+}
+```
+
+Tasks spawn onto a `ThreadPool` will, by default, spawn any subtasks onto the
+same executor.
+
+### Single-threaded execution: `LocalPool`
+
+The `LocalPool` API, on the other hand, is a bit more subtle. Here, we're
+replacing `current_thread` from [Tokio Reform] with a slightly different, more
+flexible design that integrates with the default executor system.
+
+First, we have the basic type definition and executor definition, which is much
+like `ThreadPool`:
+
+```rust
+// Note: not `Send` or `Sync`
+struct LocalPool { ... }
+impl Spawn for LocalPool { .. }
+
+impl LocalPool {
+    // create a new single-threaded executor
+    fn new() -> LocalPool;
+}
+```
+
+However, the rest of the API is more interesting:
+
+```rust
+impl LocalPool {
+    // runs the executor until `f` is resolved, spawning subtasks onto `spawn`
+    fn run_until<F, S>(&self, f: F, spawn: S) -> Result<F::Item, F::Error>
+        where F: Future, S: Spawn;
+
+    // a future that resolves when *all* spawned futures have resolved
+    fn all_done(&self) -> impl Future<Item = (), Error = ()>;
+
+    // spawns a possibly non-Send future, possible due to single-threaded execution.
+    fn spawn_local<F>(&self, F)
```

Comment:
Is there a way for a future currently being run on a `LocalPool` to be able to spawn `!Send` futures on the current `LocalPool` but futures that **are** send on the `spawn` set to context?

Reply:
What does 'spawn set to context' mean? I know contexts have to know which executor is currently required to be notified, but that behavior isn't specified by this document. Should it?

Reply:
Unrelated comment: I assume that the `Spawn` trait for `LocalPool` is as easy as:

```rust
impl Spawn for LocalPool {
    fn spawn(&self, f: Box<Future<Item = (), Error = ()> + Send>) {
        self.spawn_local(f);
    }
}
```

I.e. `spawn_local` is a superset of the requirements of `Spawn::spawn`.

Reply:
@carllerche Does `pool.run_until(my_fut, &cpu_pool)` look like what you had in mind? This would be possible with the `run_until` API specified above, although as I commented I think there should be a method which defaults to spawning onto the `LocalPool` itself.

Reply:
I don't see how `run_until` can do what I think he's asking. Here's code (cause code makes the world better, or something):

```rust
struct MixedFuture<L, P> {
    local: L,
    pool: P,
}

impl<L, P, LT, PT, E> Future for MixedFuture<L, P>
where L: Future<Item = LT, Error = E>, P: Future<Item = PT, Error = E> + Send {
    type Item = (LT, PT);
    type Error = E;
    fn poll(&mut self, ctx: &mut task::Context) -> Poll<(LT, PT), E> {
        // What should this poll function do? And which spawn is associated with ctx when it's passed here?
    }
}
```

Reply:
```rust
struct SpawnBoth<L, P> {
    future_to_spawn_on_local_pool: Option<L>,
    future_to_spawn_on_ctx_pool: Option<P>,
    local_pool_ref: Option<Rc<LocalPool>>,
}

impl<L, P> Future for SpawnBoth<L, P>
    where L: Future, P: Future + Send
{
    type Item = ();
    type Error = ();
    fn poll(&mut self, ctx: &mut task::Context) -> Poll<(), E> {
        match (
            self.future_to_spawn_on_local_pool.take(), 
            self.future_to_spawn_on_ctx_pool.take(),
            self.local_pool_ref.take(),
        ) {
            (
                Some(future_to_spawn_on_local_pool),
                Some(future_to_spawn_on_ctx_pool),
                Some(local_pool_ref),
            ) => {
                local_pool_ref.spawn(future_to_spawn_on_local_pool);
                ctx.spawn(future_to_spawn_on_ctx_pool);
            }
            _ => { panic!("polled SpawnBoth after Completion"); }
        }
        Async::Ready(())
    }
}
```